// vscode/dist/vscode/vs/workbench/browser/parts/notifications/notificationsAlerts.js
import { alert } from 'monaco-editor/esm/vs/base/browser/ui/aria/aria.js';
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { Disposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { toErrorMessage } from 'monaco-editor/esm/vs/base/common/errorMessage.js';
import { Severity, NotificationPriority } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
import { Event } from 'monaco-editor/esm/vs/base/common/event.js';
class NotificationsAlerts extends Disposable {
    constructor(model) {
        super();
        this.model = model;
        for (const notification of model.notifications) {
            this.triggerAriaAlert(notification);
        }
        this.registerListeners();
    }
    registerListeners() {
        this._register(this.model.onDidChangeNotification(e => this.onDidChangeNotification(e)));
    }
    onDidChangeNotification(e) {
        if (e.kind === 0 ) {
            this.triggerAriaAlert(e.item);
            if (e.item.severity === Severity.Error) {
                if (e.item.message.original instanceof Error) {
                    console.error(e.item.message.original);
                }
                else {
                    console.error(toErrorMessage(( e.item.message.linkedText.toString()), true));
                }
            }
        }
    }
    triggerAriaAlert(notification) {
        if (notification.priority === NotificationPriority.SILENT) {
            return;
        }
        const listener = notification.onDidChangeContent(e => {
            if (e.kind === 1 ) {
                this.doTriggerAriaAlert(notification);
            }
        });
        Event.once(notification.onDidClose)(() => listener.dispose());
        this.doTriggerAriaAlert(notification);
    }
    doTriggerAriaAlert(notification) {
        let alertText;
        if (notification.severity === Severity.Error) {
            alertText = ( localize('alertErrorMessage', "Error: {0}", ( notification.message.linkedText.toString())));
        }
        else if (notification.severity === Severity.Warning) {
            alertText = ( localize('alertWarningMessage', "Warning: {0}", ( notification.message.linkedText.toString())));
        }
        else {
            alertText = ( localize('alertInfoMessage', "Info: {0}", ( notification.message.linkedText.toString())));
        }
        alert(alertText);
    }
}
export { NotificationsAlerts };
