// vscode/dist/vscode/vs/workbench/contrib/terminal/browser/terminalConfigHelper.js
import { __decorate, __param } from '../../../../../../node_modules/tslib/tslib.es6.js';
import * as nls from 'monaco-editor/esm/vs/nls.js';
import { EDITOR_FONT_DEFAULTS } from 'monaco-editor/esm/vs/editor/common/config/editorOptions.js';
import { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
import { TERMINAL_CONFIG_SECTION, DEFAULT_FONT_WEIGHT, DEFAULT_BOLD_FONT_WEIGHT, MINIMUM_FONT_WEIGHT, MAXIMUM_FONT_WEIGHT, MINIMUM_LETTER_SPACING, DEFAULT_LETTER_SPACING, DEFAULT_LINE_HEIGHT } from '../common/terminal.js';
import Severity from 'monaco-editor/esm/vs/base/common/severity.js';
import { NeverShowAgainScope, INotificationService } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
import { Emitter } from 'monaco-editor/esm/vs/base/common/event.js';
import { basename } from 'monaco-editor/esm/vs/base/common/path.js';
import { IExtensionManagementService } from '../../../../platform/extensionManagement/common/extensionManagement.js';
import { IInstantiationService } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
import { InstallRecommendedExtensionAction } from '../../extensions/browser/extensionsActions.js';
import { IProductService } from 'monaco-editor/esm/vs/platform/product/common/productService.js';
import { isLinux, isWindows } from 'monaco-editor/esm/vs/base/common/platform.js';
let TerminalConfigHelper = class TerminalConfigHelper {
    get onConfigChanged() { return this._onConfigChanged.event; }
    constructor(_configurationService, _extensionManagementService, _notificationService, _instantiationService, _productService) {
        this._configurationService = _configurationService;
        this._extensionManagementService = _extensionManagementService;
        this._notificationService = _notificationService;
        this._instantiationService = _instantiationService;
        this._productService = _productService;
        this._linuxDistro = 1 ;
        this._onConfigChanged = ( new Emitter());
        this._recommendationsShown = false;
        this._updateConfig();
        this._configurationService.onDidChangeConfiguration(e => {
            if (e.affectsConfiguration(TERMINAL_CONFIG_SECTION)) {
                this._updateConfig();
            }
        });
        if (isLinux) {
            if (navigator.userAgent.includes('Ubuntu')) {
                this._linuxDistro = 3 ;
            }
            else if (navigator.userAgent.includes('Fedora')) {
                this._linuxDistro = 2 ;
            }
        }
    }
    _updateConfig() {
        const configValues = this._configurationService.getValue(TERMINAL_CONFIG_SECTION);
        configValues.fontWeight = this._normalizeFontWeight(configValues.fontWeight, DEFAULT_FONT_WEIGHT);
        configValues.fontWeightBold = this._normalizeFontWeight(configValues.fontWeightBold, DEFAULT_BOLD_FONT_WEIGHT);
        this.config = configValues;
        this._onConfigChanged.fire();
    }
    configFontIsMonospace() {
        const fontSize = 15;
        const fontFamily = this.config.fontFamily || this._configurationService.getValue('editor').fontFamily || EDITOR_FONT_DEFAULTS.fontFamily;
        const iRect = this._getBoundingRectFor('i', fontFamily, fontSize);
        const wRect = this._getBoundingRectFor('w', fontFamily, fontSize);
        if (!iRect || !wRect || !iRect.width || !wRect.width) {
            return true;
        }
        return iRect.width === wRect.width;
    }
    _createCharMeasureElementIfNecessary() {
        if (!this.panelContainer) {
            throw new Error('Cannot measure element when terminal is not attached');
        }
        if (!this._charMeasureElement || !this._charMeasureElement.parentElement) {
            this._charMeasureElement = document.createElement('div');
            this.panelContainer.appendChild(this._charMeasureElement);
        }
        return this._charMeasureElement;
    }
    _getBoundingRectFor(char, fontFamily, fontSize) {
        let charMeasureElement;
        try {
            charMeasureElement = this._createCharMeasureElementIfNecessary();
        }
        catch {
            return undefined;
        }
        const style = charMeasureElement.style;
        style.display = 'inline-block';
        style.fontFamily = fontFamily;
        style.fontSize = fontSize + 'px';
        style.lineHeight = 'normal';
        charMeasureElement.innerText = char;
        const rect = charMeasureElement.getBoundingClientRect();
        style.display = 'none';
        return rect;
    }
    _measureFont(fontFamily, fontSize, letterSpacing, lineHeight) {
        const rect = this._getBoundingRectFor('X', fontFamily, fontSize);
        if (this._lastFontMeasurement && (!rect || !rect.width || !rect.height)) {
            return this._lastFontMeasurement;
        }
        this._lastFontMeasurement = {
            fontFamily,
            fontSize,
            letterSpacing,
            lineHeight,
            charWidth: 0,
            charHeight: 0
        };
        if (rect && rect.width && rect.height) {
            this._lastFontMeasurement.charHeight = Math.ceil(rect.height);
            if (this.config.gpuAcceleration === 'off') {
                this._lastFontMeasurement.charWidth = rect.width;
            }
            else {
                const deviceCharWidth = Math.floor(rect.width * window.devicePixelRatio);
                const deviceCellWidth = deviceCharWidth + Math.round(letterSpacing);
                const cssCellWidth = deviceCellWidth / window.devicePixelRatio;
                this._lastFontMeasurement.charWidth = cssCellWidth - Math.round(letterSpacing) / window.devicePixelRatio;
            }
        }
        return this._lastFontMeasurement;
    }
    getFont(xtermCore, excludeDimensions) {
        const editorConfig = this._configurationService.getValue('editor');
        let fontFamily = this.config.fontFamily || editorConfig.fontFamily || EDITOR_FONT_DEFAULTS.fontFamily;
        let fontSize = this._clampInt(this.config.fontSize, 6 , 100 , EDITOR_FONT_DEFAULTS.fontSize);
        if (!this.config.fontFamily) {
            if (this._linuxDistro === 2 ) {
                fontFamily = '\'DejaVu Sans Mono\'';
            }
            if (this._linuxDistro === 3 ) {
                fontFamily = '\'Ubuntu Mono\'';
                fontSize = this._clampInt(fontSize + 2, 6 , 100 , EDITOR_FONT_DEFAULTS.fontSize);
            }
        }
        fontFamily += ', monospace';
        const letterSpacing = this.config.letterSpacing ? Math.max(Math.floor(this.config.letterSpacing), MINIMUM_LETTER_SPACING) : DEFAULT_LETTER_SPACING;
        const lineHeight = this.config.lineHeight ? Math.max(this.config.lineHeight, 1) : DEFAULT_LINE_HEIGHT;
        if (excludeDimensions) {
            return {
                fontFamily,
                fontSize,
                letterSpacing,
                lineHeight
            };
        }
        if (xtermCore) {
            if (xtermCore._renderService && xtermCore._renderService.dimensions?.css.cell.width && xtermCore._renderService.dimensions?.css.cell.height) {
                return {
                    fontFamily,
                    fontSize,
                    letterSpacing,
                    lineHeight,
                    charHeight: xtermCore._renderService.dimensions.css.cell.height / lineHeight,
                    charWidth: xtermCore._renderService.dimensions.css.cell.width - Math.round(letterSpacing) / window.devicePixelRatio
                };
            }
        }
        return this._measureFont(fontFamily, fontSize, letterSpacing, lineHeight);
    }
    _clampInt(source, minimum, maximum, fallback) {
        let r = parseInt(source, 10);
        if (isNaN(r)) {
            return fallback;
        }
        if (typeof minimum === 'number') {
            r = Math.max(minimum, r);
        }
        if (typeof maximum === 'number') {
            r = Math.min(maximum, r);
        }
        return r;
    }
    async showRecommendations(shellLaunchConfig) {
        if (this._recommendationsShown) {
            return;
        }
        this._recommendationsShown = true;
        if (isWindows && shellLaunchConfig.executable && basename(shellLaunchConfig.executable).toLowerCase() === 'wsl.exe') {
            const exeBasedExtensionTips = this._productService.exeBasedExtensionTips;
            if (!exeBasedExtensionTips || !exeBasedExtensionTips.wsl) {
                return;
            }
            const extId = ( Object.keys(exeBasedExtensionTips.wsl.recommendations)).find(extId => exeBasedExtensionTips.wsl.recommendations[extId].important);
            if (extId && !(await this._isExtensionInstalled(extId))) {
                this._notificationService.prompt(Severity.Info, ( nls.localize(
                    'useWslExtension.title',
                    "The '{0}' extension is recommended for opening a terminal in WSL.",
                    exeBasedExtensionTips.wsl.friendlyName
                )), [
                    {
                        label: ( nls.localize('install', 'Install')),
                        run: () => {
                            this._instantiationService.createInstance(InstallRecommendedExtensionAction, extId).run();
                        }
                    }
                ], {
                    sticky: true,
                    neverShowAgain: { id: 'terminalConfigHelper/launchRecommendationsIgnore', scope: NeverShowAgainScope.APPLICATION },
                    onCancel: () => { }
                });
            }
        }
    }
    async _isExtensionInstalled(id) {
        const extensions = await this._extensionManagementService.getInstalled();
        return ( extensions.some(e => e.identifier.id === id));
    }
    _normalizeFontWeight(input, defaultWeight) {
        if (input === 'normal' || input === 'bold') {
            return input;
        }
        return this._clampInt(input, MINIMUM_FONT_WEIGHT, MAXIMUM_FONT_WEIGHT, defaultWeight);
    }
};
TerminalConfigHelper = ( __decorate([
    ( __param(0, IConfigurationService)),
    ( __param(1, IExtensionManagementService)),
    ( __param(2, INotificationService)),
    ( __param(3, IInstantiationService)),
    ( __param(4, IProductService))
], TerminalConfigHelper));
export { TerminalConfigHelper };
