// vscode/dist/vscode/vs/workbench/browser/parts/editor/binaryEditor.js
import { __decorate, __param } from '../../../../../../node_modules/tslib/tslib.es6.js';
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { Emitter } from 'monaco-editor/esm/vs/base/common/event.js';
import { BinaryEditorModel } from '../../../common/editor/binaryEditorModel.js';
import { IStorageService } from 'monaco-editor/esm/vs/platform/storage/common/storage.js';
import { ByteSize } from 'monaco-editor/esm/vs/platform/files/common/files.js';
import { EditorPlaceholder } from './editorPlaceholder.js';
let BaseBinaryResourceEditor = class BaseBinaryResourceEditor extends EditorPlaceholder {
    constructor(id, callbacks, telemetryService, themeService, storageService) {
        super(id, telemetryService, themeService, storageService);
        this.callbacks = callbacks;
        this._onDidChangeMetadata = this._register(( new Emitter()));
        this.onDidChangeMetadata = this._onDidChangeMetadata.event;
        this._onDidOpenInPlace = this._register(( new Emitter()));
        this.onDidOpenInPlace = this._onDidOpenInPlace.event;
    }
    getTitle() {
        return this.input ? this.input.getName() : ( localize('binaryEditor', "Binary Viewer"));
    }
    async getContents(input, options) {
        const model = await input.resolve(options);
        if (!(model instanceof BinaryEditorModel)) {
            throw new Error('Unable to open file as binary');
        }
        const size = model.getSize();
        this.handleMetadataChanged(typeof size === 'number' ? ByteSize.formatSize(size) : '');
        return {
            icon: '$(warning)',
            label: ( localize(
                'binaryError',
                "The file is not displayed in the text editor because it is either binary or uses an unsupported text encoding."
            )),
            actions: [
                {
                    label: ( localize('openAnyway', "Open Anyway")),
                    run: async () => {
                        await this.callbacks.openInternal(input, options);
                        this._onDidOpenInPlace.fire();
                    }
                }
            ]
        };
    }
    handleMetadataChanged(meta) {
        this.metadata = meta;
        this._onDidChangeMetadata.fire();
    }
    getMetadata() {
        return this.metadata;
    }
};
BaseBinaryResourceEditor = ( __decorate([
    ( __param(4, IStorageService))
], BaseBinaryResourceEditor));
export { BaseBinaryResourceEditor };
