// vscode/dist/vscode/vs/workbench/browser/parts/editor/editorGroupWatermark.js
import { __decorate, __param } from '../../../../../../node_modules/tslib/tslib.es6.js';
import { Disposable, DisposableStore } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { isMacintosh, isWeb } from 'monaco-editor/esm/vs/base/common/platform.js';
import { IKeybindingService } from 'monaco-editor/esm/vs/platform/keybinding/common/keybinding.js';
import { ITelemetryService } from 'monaco-editor/esm/vs/platform/telemetry/common/telemetry.js';
import * as nls from 'monaco-editor/esm/vs/nls.js';
import { IWorkspaceContextService } from 'monaco-editor/esm/vs/platform/workspace/common/workspace.js';
import { ILifecycleService } from '../../../services/lifecycle/common/lifecycle.js';
import { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
import { h, append, $, clearNode } from 'monaco-editor/esm/vs/base/browser/dom.js';
import { CommandsRegistry } from 'monaco-editor/esm/vs/platform/commands/common/commands.js';
import { IContextKeyService, ContextKeyExpr } from 'monaco-editor/esm/vs/platform/contextkey/common/contextkey.js';
const showCommands = { text: ( nls.localize('watermark.showCommands', "Show All Commands")), id: 'workbench.action.showCommands' };
const quickAccess = { text: ( nls.localize('watermark.quickAccess', "Go to File")), id: 'workbench.action.quickOpen' };
const openFileNonMacOnly = { text: ( nls.localize('watermark.openFile', "Open File")), id: 'workbench.action.files.openFile', mac: false };
const openFolderNonMacOnly = { text: ( nls.localize('watermark.openFolder', "Open Folder")), id: 'workbench.action.files.openFolder', mac: false };
const openFileOrFolderMacOnly = { text: ( nls.localize('watermark.openFileFolder', "Open File or Folder")), id: 'workbench.action.files.openFileFolder', mac: true };
const openRecent = { text: ( nls.localize('watermark.openRecent', "Open Recent")), id: 'workbench.action.openRecent' };
const newUntitledFile = { text: ( nls.localize('watermark.newUntitledFile', "New Untitled Text File")), id: 'workbench.action.files.newUntitledFile' };
const newUntitledFileMacOnly = Object.assign({ mac: true }, newUntitledFile);
const findInFiles = { text: ( nls.localize('watermark.findInFiles', "Find in Files")), id: 'workbench.action.findInFiles' };
const toggleTerminal = { text: ( nls.localize(
    { key: 'watermark.toggleTerminal', comment: ['toggle is a verb here'] },
    "Toggle Terminal"
)), id: 'workbench.action.terminal.toggleTerminal', when: ( ContextKeyExpr.equals('terminalProcessSupported', true)) };
const startDebugging = { text: ( nls.localize('watermark.startDebugging', "Start Debugging")), id: 'workbench.action.debug.start', when: ( ContextKeyExpr.equals('terminalProcessSupported', true)) };
const toggleFullscreen = { text: ( nls.localize(
    { key: 'watermark.toggleFullscreen', comment: ['toggle is a verb here'] },
    "Toggle Full Screen"
)), id: 'workbench.action.toggleFullScreen', when: ( ( ContextKeyExpr.equals('terminalProcessSupported', true)).negate()) };
const showSettings = { text: ( nls.localize('watermark.showSettings', "Show Settings")), id: 'workbench.action.openSettings', when: ( ( ContextKeyExpr.equals('terminalProcessSupported', true)).negate()) };
const noFolderEntries = [
    showCommands,
    openFileNonMacOnly,
    openFolderNonMacOnly,
    openFileOrFolderMacOnly,
    openRecent,
    newUntitledFileMacOnly
];
const folderEntries = [
    showCommands,
    quickAccess,
    findInFiles,
    startDebugging,
    toggleTerminal,
    toggleFullscreen,
    showSettings
];
let EditorGroupWatermark = class EditorGroupWatermark extends Disposable {
    constructor(container, lifecycleService, keybindingService, contextService, contextKeyService, configurationService, telemetryService) {
        super();
        this.lifecycleService = lifecycleService;
        this.keybindingService = keybindingService;
        this.contextService = contextService;
        this.contextKeyService = contextKeyService;
        this.configurationService = configurationService;
        this.telemetryService = telemetryService;
        this.transientDisposables = this._register(( new DisposableStore()));
        this.enabled = false;
        const elements = h('.editor-group-watermark', [
            h('.letterpress'),
            h('.shortcuts@shortcuts'),
        ]);
        append(container, elements.root);
        this.shortcuts = elements.shortcuts;
        this.registerListeners();
        this.workbenchState = contextService.getWorkbenchState();
        this.render();
    }
    registerListeners() {
        this.lifecycleService.onDidShutdown(() => this.dispose());
        this._register(this.configurationService.onDidChangeConfiguration(e => {
            if (e.affectsConfiguration('workbench.tips.enabled')) {
                this.render();
            }
        }));
        this._register(this.contextService.onDidChangeWorkbenchState(workbenchState => {
            if (this.workbenchState === workbenchState) {
                return;
            }
            this.workbenchState = workbenchState;
            this.render();
        }));
        const allEntriesWhenClauses = ( [...noFolderEntries, ...folderEntries].filter(entry => entry.when !== undefined).map(entry => entry.when));
        const allKeys = ( new Set());
        allEntriesWhenClauses.forEach(when => ( when.keys()).forEach(key => allKeys.add(key)));
        this._register(this.contextKeyService.onDidChangeContext(e => {
            if (e.affectsSome(allKeys)) {
                this.render();
            }
        }));
    }
    render() {
        const enabled = this.configurationService.getValue('workbench.tips.enabled');
        if (enabled === this.enabled) {
            return;
        }
        this.enabled = enabled;
        this.clear();
        if (!enabled) {
            return;
        }
        const box = append(this.shortcuts, $('.watermark-box'));
        const folder = this.workbenchState !== 1 ;
        (folder ? folderEntries : noFolderEntries)
            .filter(entry => !('when' in entry) || this.contextKeyService.contextMatchesRules(entry.when))
            .filter(entry => !('mac' in entry) || entry.mac === (isMacintosh && !isWeb))
            .filter(entry => !!CommandsRegistry.getCommand(entry.id));
        const update = () => {
            clearNode(box);
        };
        update();
        this.transientDisposables.add(this.keybindingService.onDidUpdateKeybindings(update));
        this.telemetryService.publicLog('watermark:open');
    }
    clear() {
        clearNode(this.shortcuts);
        this.transientDisposables.clear();
    }
    dispose() {
        super.dispose();
        this.clear();
    }
};
EditorGroupWatermark = ( __decorate([
    ( __param(1, ILifecycleService)),
    ( __param(2, IKeybindingService)),
    ( __param(3, IWorkspaceContextService)),
    ( __param(4, IContextKeyService)),
    ( __param(5, IConfigurationService)),
    ( __param(6, ITelemetryService))
], EditorGroupWatermark));
export { EditorGroupWatermark };
