// vscode/dist/vscode/vs/workbench/contrib/debug/common/debugContentProvider.js
import { __decorate, __param } from '../../../../../../node_modules/tslib/tslib.es6.js';
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { getMimeTypes } from 'monaco-editor/esm/vs/editor/common/services/languagesAssociations.js';
import { IModelService } from 'monaco-editor/esm/vs/editor/common/services/model.js';
import { ILanguageService } from 'monaco-editor/esm/vs/editor/common/languages/language.js';
import { ITextModelService } from 'monaco-editor/esm/vs/editor/common/services/resolverService.js';
import { DEBUG_SCHEME, IDebugService } from './debug.js';
import { Source } from './debugSource.js';
import { IEditorWorkerService } from 'monaco-editor/esm/vs/editor/common/services/editorWorker.js';
import { EditOperation } from 'monaco-editor/esm/vs/editor/common/core/editOperation.js';
import { Range } from 'monaco-editor/esm/vs/editor/common/core/range.js';
import { CancellationTokenSource } from 'monaco-editor/esm/vs/base/common/cancellation.js';
import { PLAINTEXT_LANGUAGE_ID } from 'monaco-editor/esm/vs/editor/common/languages/modesRegistry.js';
import { ErrorNoTelemetry } from 'monaco-editor/esm/vs/base/common/errors.js';
let DebugContentProvider = class DebugContentProvider {
    constructor(textModelResolverService, debugService, modelService, languageService, editorWorkerService) {
        this.debugService = debugService;
        this.modelService = modelService;
        this.languageService = languageService;
        this.editorWorkerService = editorWorkerService;
        this.pendingUpdates = ( new Map());
        textModelResolverService.registerTextModelContentProvider(DEBUG_SCHEME, this);
        DebugContentProvider.INSTANCE = this;
    }
    dispose() {
        this.pendingUpdates.forEach(cancellationSource => cancellationSource.dispose());
    }
    provideTextContent(resource) {
        return this.createOrUpdateContentModel(resource, true);
    }
    static refreshDebugContent(resource) {
        DebugContentProvider.INSTANCE?.createOrUpdateContentModel(resource, false);
    }
    createOrUpdateContentModel(resource, createIfNotExists) {
        const model = this.modelService.getModel(resource);
        if (!model && !createIfNotExists) {
            return null;
        }
        let session;
        if (resource.query) {
            const data = Source.getEncodedDebugData(resource);
            session = this.debugService.getModel().getSession(data.sessionId);
        }
        if (!session) {
            session = this.debugService.getViewModel().focusedSession;
        }
        if (!session) {
            return Promise.reject(( new ErrorNoTelemetry(( localize('unable', "Unable to resolve the resource without a debug session")))));
        }
        const createErrModel = (errMsg) => {
            this.debugService.sourceIsNotAvailable(resource);
            const languageSelection = this.languageService.createById(PLAINTEXT_LANGUAGE_ID);
            const message = errMsg
                ? ( localize(
                'canNotResolveSourceWithError',
                "Could not load source '{0}': {1}.",
                resource.path,
                errMsg
            ))
                : ( localize('canNotResolveSource', "Could not load source '{0}'.", resource.path));
            return this.modelService.createModel(message, languageSelection, resource);
        };
        return session.loadSource(resource).then(response => {
            if (response && response.body) {
                if (model) {
                    const newContent = response.body.content;
                    const cancellationSource = this.pendingUpdates.get(model.id);
                    cancellationSource?.cancel();
                    const myToken = ( new CancellationTokenSource());
                    this.pendingUpdates.set(model.id, myToken);
                    return this.editorWorkerService.computeMoreMinimalEdits(model.uri, [{ text: newContent, range: model.getFullModelRange() }]).then(edits => {
                        this.pendingUpdates.delete(model.id);
                        if (!myToken.token.isCancellationRequested && edits && edits.length > 0) {
                            model.applyEdits(( edits.map(edit => EditOperation.replace(Range.lift(edit.range), edit.text))));
                        }
                        return model;
                    });
                }
                else {
                    const mime = response.body.mimeType || getMimeTypes(resource)[0];
                    const languageSelection = this.languageService.createByMimeType(mime);
                    return this.modelService.createModel(response.body.content, languageSelection, resource);
                }
            }
            return createErrModel();
        }, (err) => createErrModel(err.message));
    }
};
DebugContentProvider = ( __decorate([
    ( __param(0, ITextModelService)),
    ( __param(1, IDebugService)),
    ( __param(2, IModelService)),
    ( __param(3, ILanguageService)),
    ( __param(4, IEditorWorkerService))
], DebugContentProvider));
export { DebugContentProvider };
