// vscode/dist/vscode/vs/workbench/browser/parts/paneCompositePart.js
import { __decorate, __param } from '../../../../../node_modules/tslib/tslib.es6.js';
import { Event } from 'monaco-editor/esm/vs/base/common/event.js';
import { assertIsDefined } from 'monaco-editor/esm/vs/base/common/types.js';
import { IInstantiationService } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
import { ActivitybarPart } from './activitybar/activitybarPart.js';
import { AuxiliaryBarPart } from './auxiliarybar/auxiliaryBarPart.js';
import { PanelPart } from './panel/panelPart.js';
import { SidebarPart } from './sidebar/sidebarPart.js';
import { ViewContainerLocations } from '../../common/views.js';
import { Disposable, DisposableStore } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
let PaneCompositeParts = class PaneCompositeParts extends Disposable {
    constructor(instantiationService) {
        super();
        this.paneCompositeParts = ( new Map());
        this.paneCompositeSelectorParts = ( new Map());
        const panelPart = instantiationService.createInstance(PanelPart);
        const sideBarPart = instantiationService.createInstance(SidebarPart);
        const auxiliaryBarPart = instantiationService.createInstance(AuxiliaryBarPart);
        const activityBarPart = instantiationService.createInstance(ActivitybarPart, sideBarPart);
        this.paneCompositeParts.set(1 , panelPart);
        this.paneCompositeParts.set(0 , sideBarPart);
        this.paneCompositeParts.set(2 , auxiliaryBarPart);
        this.paneCompositeSelectorParts.set(1 , panelPart);
        this.paneCompositeSelectorParts.set(0 , activityBarPart);
        this.paneCompositeSelectorParts.set(2 , auxiliaryBarPart);
        const eventDisposables = this._register(( new DisposableStore()));
        this.onDidPaneCompositeOpen = Event.any(...( ViewContainerLocations.map(loc => ( Event.map(
            this.paneCompositeParts.get(loc).onDidPaneCompositeOpen,
            composite => { return { composite, viewContainerLocation: loc }; },
            eventDisposables
        )))));
        this.onDidPaneCompositeClose = Event.any(...( ViewContainerLocations.map(loc => ( Event.map(
            this.paneCompositeParts.get(loc).onDidPaneCompositeClose,
            composite => { return { composite, viewContainerLocation: loc }; },
            eventDisposables
        )))));
    }
    openPaneComposite(id, viewContainerLocation, focus) {
        return this.getPartByLocation(viewContainerLocation).openPaneComposite(id, focus);
    }
    getActivePaneComposite(viewContainerLocation) {
        return this.getPartByLocation(viewContainerLocation).getActivePaneComposite();
    }
    getPaneComposite(id, viewContainerLocation) {
        return this.getPartByLocation(viewContainerLocation).getPaneComposite(id);
    }
    getPaneComposites(viewContainerLocation) {
        return this.getPartByLocation(viewContainerLocation).getPaneComposites();
    }
    getPinnedPaneCompositeIds(viewContainerLocation) {
        return this.getSelectorPartByLocation(viewContainerLocation).getPinnedPaneCompositeIds();
    }
    getVisiblePaneCompositeIds(viewContainerLocation) {
        return this.getSelectorPartByLocation(viewContainerLocation).getVisiblePaneCompositeIds();
    }
    getProgressIndicator(id, viewContainerLocation) {
        return this.getPartByLocation(viewContainerLocation).getProgressIndicator(id);
    }
    hideActivePaneComposite(viewContainerLocation) {
        this.getPartByLocation(viewContainerLocation).hideActivePaneComposite();
    }
    getLastActivePaneCompositeId(viewContainerLocation) {
        return this.getPartByLocation(viewContainerLocation).getLastActivePaneCompositeId();
    }
    showActivity(id, viewContainerLocation, badge, clazz, priority) {
        return this.getSelectorPartByLocation(viewContainerLocation).showActivity(id, badge, clazz, priority);
    }
    getPartByLocation(viewContainerLocation) {
        return assertIsDefined(this.paneCompositeParts.get(viewContainerLocation));
    }
    getSelectorPartByLocation(viewContainerLocation) {
        return assertIsDefined(this.paneCompositeSelectorParts.get(viewContainerLocation));
    }
};
PaneCompositeParts = ( __decorate([
    ( __param(0, IInstantiationService))
], PaneCompositeParts));
export { PaneCompositeParts };
