// vscode/dist/api.js
import { Breakpoint as Breakpoint$1, CallHierarchyIncomingCall as CallHierarchyIncomingCall$1, CallHierarchyItem as CallHierarchyItem$1, CallHierarchyOutgoingCall as CallHierarchyOutgoingCall$1, CodeAction as CodeAction$1, CodeActionKind as CodeActionKind$1, CodeActionTriggerKind as CodeActionTriggerKind$1, CodeLens as CodeLens$1, Color as Color$1, ColorInformation as ColorInformation$1, ColorPresentation as ColorPresentation$1, ColorThemeKind as ColorThemeKind$1, CommentMode as CommentMode$1, CommentThreadCollapsibleState as CommentThreadCollapsibleState$1, CompletionItem as CompletionItem$1, CompletionItemKind as CompletionItemKind$1, CompletionItemTag as CompletionItemTag$1, CompletionList as CompletionList$1, CompletionTriggerKind as CompletionTriggerKind$1, ConfigurationTarget as ConfigurationTarget$1, CustomExecution as CustomExecution$1, DebugAdapterExecutable as DebugAdapterExecutable$1, DebugAdapterInlineImplementation as DebugAdapterInlineImplementation$1, DebugAdapterNamedPipeServer as DebugAdapterNamedPipeServer$1, DebugAdapterServer as DebugAdapterServer$1, DebugConsoleMode as DebugConsoleMode$1, DecorationRangeBehavior as DecorationRangeBehavior$1, Diagnostic as Diagnostic$1, DiagnosticRelatedInformation as DiagnosticRelatedInformation$1, DiagnosticSeverity as DiagnosticSeverity$1, DiagnosticTag as DiagnosticTag$1, Disposable as Disposable$1, DocumentHighlight as DocumentHighlight$1, DocumentHighlightKind as DocumentHighlightKind$1, DocumentLink as DocumentLink$1, DocumentSymbol as DocumentSymbol$1, EndOfLine as EndOfLine$1, EnvironmentVariableMutatorType as EnvironmentVariableMutatorType$1, EvaluatableExpression as EvaluatableExpression$1, InlineValueText as InlineValueText$1, InlineValueVariableLookup as InlineValueVariableLookup$1, InlineValueEvaluatableExpression as InlineValueEvaluatableExpression$1, ExtensionKind as ExtensionKind$1, ExtensionMode as ExtensionMode$1, FileChangeType as FileChangeType$1, FileDecoration as FileDecoration$1, FileSystemError as FileSystemError$1, FoldingRange as FoldingRange$1, FoldingRangeKind as FoldingRangeKind$1, FunctionBreakpoint as FunctionBreakpoint$1, Hover as Hover$1, Location as Location$1, MarkdownString as MarkdownString$1, ParameterInformation as ParameterInformation$1, Position as Position$1, ProcessExecution as ProcessExecution$1, ProgressLocation as ProgressLocation$1, QuickInputButtons as QuickInputButtons$1, Range as Range$1, RelativePattern as RelativePattern$1, Selection as Selection$1, SelectionRange as SelectionRange$1, SemanticTokens as SemanticTokens$1, SemanticTokensBuilder as SemanticTokensBuilder$1, SemanticTokensEdit as SemanticTokensEdit$1, SemanticTokensEdits as SemanticTokensEdits$1, SemanticTokensLegend as SemanticTokensLegend$1, ShellExecution as ShellExecution$1, ShellQuoting as ShellQuoting$1, SignatureHelp as SignatureHelp$1, SignatureHelpTriggerKind as SignatureHelpTriggerKind$1, SignatureInformation as SignatureInformation$1, SnippetString as SnippetString$1, SourceBreakpoint as SourceBreakpoint$1, StatusBarAlignment as StatusBarAlignment$1, SymbolInformation as SymbolInformation$1, SymbolKind as SymbolKind$1, SymbolTag as SymbolTag$1, TerminalLink as TerminalLink$1, TerminalLocation as TerminalLocation$1, TerminalProfile as TerminalProfile$1, TextDocumentSaveReason as TextDocumentSaveReason$1, TextEdit as TextEdit$1, TextEditorLineNumbersStyle as TextEditorLineNumbersStyle$1, TextEditorRevealType as TextEditorRevealType$1, TextEditorSelectionChangeKind as TextEditorSelectionChangeKind$1, TextDocumentChangeReason as TextDocumentChangeReason$1, ThemeColor as ThemeColor$1, ThemeIcon as ThemeIcon$1, TreeItem as TreeItem$1, TreeItemCollapsibleState as TreeItemCollapsibleState$1, TypeHierarchyItem as TypeHierarchyItem$1, ViewColumn as ViewColumn$1, WorkspaceEdit as WorkspaceEdit$1, InlayHint as InlayHint$1, InlayHintLabelPart as InlayHintLabelPart$1, InlayHintKind as InlayHintKind$1, LinkedEditingRanges as LinkedEditingRanges$1, TestRunRequest as TestRunRequest$1, TestMessage as TestMessage$1, TestTag as TestTag$1, TestRunProfileKind as TestRunProfileKind$1, DataTransfer as DataTransfer$1, DataTransferItem as DataTransferItem$1, LanguageStatusSeverity as LanguageStatusSeverity$1, QuickPickItemKind as QuickPickItemKind$1, TextTabInput, TextDiffTabInput, CustomEditorTabInput, NotebookEditorTabInput, NotebookDiffEditorTabInput, WebviewEditorTabInput, TerminalEditorTabInput, InputBoxValidationSeverity as InputBoxValidationSeverity$1, InlineSuggestionList, InlineCompletionTriggerKind as InlineCompletionTriggerKind$1, InlineSuggestion, DocumentDropEdit as DocumentDropEdit$1, SnippetTextEdit as SnippetTextEdit$1, TerminalExitReason as TerminalExitReason$1, DocumentPasteEdit as DocumentPasteEdit$1, ExternalUriOpenerPriority as ExternalUriOpenerPriority$1, TreeItemCheckboxState as TreeItemCheckboxState$1 } from './vscode/vs/workbench/api/common/extHostTypes.js';
import * as errors from 'monaco-editor/esm/vs/base/common/errors.js';
import { DebugConfigurationProviderTriggerKind as DebugConfigurationProviderTriggerKind$1 } from './vscode/vs/workbench/contrib/debug/common/debug.js';
import * as files from 'monaco-editor/esm/vs/platform/files/common/files.js';
import { UIKind as UIKind$1 } from './vscode/vs/workbench/services/extensions/common/extensionHostProtocol.js';
import * as cancellation from 'monaco-editor/esm/vs/base/common/cancellation.js';
import * as event from 'monaco-editor/esm/vs/base/common/event.js';
import * as languageConfiguration from 'monaco-editor/esm/vs/editor/common/languages/languageConfiguration.js';
import * as model from 'monaco-editor/esm/vs/editor/common/model.js';
import * as editorOptions from 'monaco-editor/esm/vs/editor/common/config/editorOptions.js';
import * as uri from 'monaco-editor/esm/vs/base/common/uri.js';
import * as log from 'monaco-editor/esm/vs/platform/log/common/log.js';
import { TelemetryTrustedValue as TelemetryTrustedValue$1 } from './vscode/vs/platform/telemetry/common/telemetryUtils.js';
import { TextSearchCompleteMessageType as TextSearchCompleteMessageType$1 } from './vscode/vs/workbench/services/search/common/searchExtTypes.js';
let defaultApi;
function setDefaultApi(api) {
    defaultApi = api;
}
function createProxy(key) {
    return new Proxy({}, {
        get(target, p) {
            if (defaultApi == null) {
                throw new Error('Default api is not ready yet, do not forget to call `initialize` from \'vscode/extensions\'');
            }
            return defaultApi[key][p];
        }
    });
}
const unsupported = undefined;
const api = {
    version: "1.81.0",
    tasks: createProxy('tasks'),
    notebooks: createProxy('notebooks'),
    scm: createProxy('scm'),
    comments: createProxy('comments'),
    authentication: createProxy('authentication'),
    tests: createProxy('tests'),
    extensions: createProxy('extensions'),
    debug: createProxy('debug'),
    env: createProxy('env'),
    commands: createProxy('commands'),
    window: createProxy('window'),
    workspace: createProxy('workspace'),
    languages: createProxy('languages'),
    l10n: createProxy('l10n'),
    Breakpoint: Breakpoint$1,
    CallHierarchyIncomingCall: CallHierarchyIncomingCall$1,
    CallHierarchyItem: CallHierarchyItem$1,
    CallHierarchyOutgoingCall: CallHierarchyOutgoingCall$1,
    CancellationError: errors.CancellationError,
    CancellationTokenSource: cancellation.CancellationTokenSource,
    CodeAction: CodeAction$1,
    CodeActionKind: CodeActionKind$1,
    CodeActionTriggerKind: CodeActionTriggerKind$1,
    CodeLens: CodeLens$1,
    Color: Color$1,
    ColorInformation: ColorInformation$1,
    ColorPresentation: ColorPresentation$1,
    ColorThemeKind: ColorThemeKind$1,
    CommentMode: CommentMode$1,
    CommentThreadCollapsibleState: CommentThreadCollapsibleState$1,
    CompletionItem: CompletionItem$1,
    CompletionItemKind: CompletionItemKind$1,
    CompletionItemTag: CompletionItemTag$1,
    CompletionList: CompletionList$1,
    CompletionTriggerKind: CompletionTriggerKind$1,
    ConfigurationTarget: ConfigurationTarget$1,
    CustomExecution: CustomExecution$1,
    DebugAdapterExecutable: DebugAdapterExecutable$1,
    DebugAdapterInlineImplementation: DebugAdapterInlineImplementation$1,
    DebugAdapterNamedPipeServer: DebugAdapterNamedPipeServer$1,
    DebugAdapterServer: DebugAdapterServer$1,
    DebugConfigurationProviderTriggerKind: DebugConfigurationProviderTriggerKind$1,
    DebugConsoleMode: DebugConsoleMode$1,
    DecorationRangeBehavior: DecorationRangeBehavior$1,
    Diagnostic: Diagnostic$1,
    DiagnosticRelatedInformation: DiagnosticRelatedInformation$1,
    DiagnosticSeverity: DiagnosticSeverity$1,
    DiagnosticTag: DiagnosticTag$1,
    Disposable: Disposable$1,
    DocumentHighlight: DocumentHighlight$1,
    DocumentHighlightKind: DocumentHighlightKind$1,
    DocumentLink: DocumentLink$1,
    DocumentSymbol: DocumentSymbol$1,
    EndOfLine: EndOfLine$1,
    EnvironmentVariableMutatorType: EnvironmentVariableMutatorType$1,
    EvaluatableExpression: EvaluatableExpression$1,
    InlineValueText: InlineValueText$1,
    InlineValueVariableLookup: InlineValueVariableLookup$1,
    InlineValueEvaluatableExpression: InlineValueEvaluatableExpression$1,
    EventEmitter: event.Emitter,
    ExtensionKind: ExtensionKind$1,
    ExtensionMode: ExtensionMode$1,
    FileChangeType: FileChangeType$1,
    FileDecoration: FileDecoration$1,
    FileSystemError: FileSystemError$1,
    FileType: files.FileType,
    FilePermission: files.FilePermission,
    FoldingRange: FoldingRange$1,
    FoldingRangeKind: FoldingRangeKind$1,
    FunctionBreakpoint: FunctionBreakpoint$1,
    Hover: Hover$1,
    IndentAction: languageConfiguration.IndentAction,
    Location: Location$1,
    MarkdownString: MarkdownString$1,
    OverviewRulerLane: model.OverviewRulerLane,
    ParameterInformation: ParameterInformation$1,
    Position: Position$1,
    ProcessExecution: ProcessExecution$1,
    ProgressLocation: ProgressLocation$1,
    QuickInputButtons: QuickInputButtons$1,
    Range: Range$1,
    RelativePattern: RelativePattern$1,
    Selection: Selection$1,
    SelectionRange: SelectionRange$1,
    SemanticTokens: SemanticTokens$1,
    SemanticTokensBuilder: SemanticTokensBuilder$1,
    SemanticTokensEdit: SemanticTokensEdit$1,
    SemanticTokensEdits: SemanticTokensEdits$1,
    SemanticTokensLegend: SemanticTokensLegend$1,
    ShellExecution: ShellExecution$1,
    ShellQuoting: ShellQuoting$1,
    SignatureHelp: SignatureHelp$1,
    SignatureHelpTriggerKind: SignatureHelpTriggerKind$1,
    SignatureInformation: SignatureInformation$1,
    SnippetString: SnippetString$1,
    SourceBreakpoint: SourceBreakpoint$1,
    StatusBarAlignment: StatusBarAlignment$1,
    SymbolInformation: SymbolInformation$1,
    SymbolKind: SymbolKind$1,
    SymbolTag: SymbolTag$1,
    Task: unsupported,
    TaskGroup: unsupported,
    TaskPanelKind: unsupported,
    TaskRevealKind: unsupported,
    TaskScope: unsupported,
    TerminalLink: TerminalLink$1,
    TerminalLocation: TerminalLocation$1,
    TerminalProfile: TerminalProfile$1,
    TextDocumentSaveReason: TextDocumentSaveReason$1,
    TextEdit: TextEdit$1,
    TextEditorCursorStyle: editorOptions.TextEditorCursorStyle,
    TextEditorLineNumbersStyle: TextEditorLineNumbersStyle$1,
    TextEditorRevealType: TextEditorRevealType$1,
    TextEditorSelectionChangeKind: TextEditorSelectionChangeKind$1,
    TextDocumentChangeReason: TextDocumentChangeReason$1,
    ThemeColor: ThemeColor$1,
    ThemeIcon: ThemeIcon$1,
    TreeItem: TreeItem$1,
    TreeItemCollapsibleState: TreeItemCollapsibleState$1,
    TypeHierarchyItem: TypeHierarchyItem$1,
    UIKind: UIKind$1,
    Uri: uri.URI,
    ViewColumn: ViewColumn$1,
    WorkspaceEdit: WorkspaceEdit$1,
    InlayHint: InlayHint$1,
    InlayHintLabelPart: InlayHintLabelPart$1,
    InlayHintKind: InlayHintKind$1,
    NotebookRange: unsupported,
    NotebookCellKind: unsupported,
    NotebookCellData: unsupported,
    NotebookData: unsupported,
    NotebookCellStatusBarAlignment: unsupported,
    NotebookCellOutput: unsupported,
    NotebookCellOutputItem: unsupported,
    NotebookCellStatusBarItem: unsupported,
    NotebookControllerAffinity: unsupported,
    LinkedEditingRanges: LinkedEditingRanges$1,
    TestRunRequest: TestRunRequest$1,
    TestMessage: TestMessage$1,
    TestTag: TestTag$1,
    TestRunProfileKind: TestRunProfileKind$1,
    DataTransfer: DataTransfer$1,
    DataTransferItem: DataTransferItem$1,
    LanguageStatusSeverity: LanguageStatusSeverity$1,
    QuickPickItemKind: QuickPickItemKind$1,
    TabInputText: TextTabInput,
    TabInputTextDiff: TextDiffTabInput,
    TabInputCustom: CustomEditorTabInput,
    TabInputNotebook: NotebookEditorTabInput,
    TabInputNotebookDiff: NotebookDiffEditorTabInput,
    TabInputWebview: WebviewEditorTabInput,
    TabInputTerminal: TerminalEditorTabInput,
    InputBoxValidationSeverity: InputBoxValidationSeverity$1,
    InlineCompletionList: InlineSuggestionList,
    InlineCompletionTriggerKind: InlineCompletionTriggerKind$1,
    InlineCompletionItem: InlineSuggestion,
    DocumentDropEdit: DocumentDropEdit$1,
    NotebookEditorRevealType: unsupported,
    SnippetTextEdit: SnippetTextEdit$1,
    NotebookEdit: unsupported,
    LogLevel: log.LogLevel,
    TerminalExitReason: TerminalExitReason$1,
    CommentThreadState: unsupported,
    TelemetryTrustedValue: TelemetryTrustedValue$1,
    TextSearchCompleteMessageType: TextSearchCompleteMessageType$1,
    DocumentPasteEdit: DocumentPasteEdit$1,
    ExternalUriOpenerPriority: ExternalUriOpenerPriority$1,
    TreeItemCheckboxState: TreeItemCheckboxState$1
};
const { version, tasks, notebooks, scm, comments, authentication, tests, extensions, debug, env, commands, window, workspace, languages, l10n, Breakpoint, CallHierarchyIncomingCall, CallHierarchyItem, CallHierarchyOutgoingCall, CancellationError, CancellationTokenSource, CodeAction, CodeActionKind, CodeActionTriggerKind, CodeLens, Color, ColorInformation, ColorPresentation, ColorThemeKind, CommentMode, CommentThreadCollapsibleState, CompletionItem, CompletionItemKind, CompletionItemTag, CompletionList, CompletionTriggerKind, ConfigurationTarget, CustomExecution, DebugAdapterExecutable, DebugAdapterInlineImplementation, DebugAdapterNamedPipeServer, DebugAdapterServer, DebugConfigurationProviderTriggerKind, DebugConsoleMode, DecorationRangeBehavior, Diagnostic, DiagnosticRelatedInformation, DiagnosticSeverity, DiagnosticTag, Disposable, DocumentHighlight, DocumentHighlightKind, DocumentLink, DocumentSymbol, EndOfLine, EnvironmentVariableMutatorType, EvaluatableExpression, InlineValueText, InlineValueVariableLookup, InlineValueEvaluatableExpression, EventEmitter, ExtensionKind, ExtensionMode, FileChangeType, FileDecoration, FileSystemError, FileType, FilePermission, FoldingRange, FoldingRangeKind, FunctionBreakpoint, Hover, IndentAction, Location, MarkdownString, OverviewRulerLane, ParameterInformation, Position, ProcessExecution, ProgressLocation, QuickInputButtons, Range, RelativePattern, Selection, SelectionRange, SemanticTokens, SemanticTokensBuilder, SemanticTokensEdit, SemanticTokensEdits, SemanticTokensLegend, ShellExecution, ShellQuoting, SignatureHelp, SignatureHelpTriggerKind, SignatureInformation, SnippetString, SourceBreakpoint, StatusBarAlignment, SymbolInformation, SymbolKind, SymbolTag, Task, TaskGroup, TaskPanelKind, TaskRevealKind, TaskScope, TerminalLink, TerminalLocation, TerminalProfile, TextDocumentSaveReason, TextEdit, TextEditorCursorStyle, TextEditorLineNumbersStyle, TextEditorRevealType, TextEditorSelectionChangeKind, TextDocumentChangeReason, ThemeColor, ThemeIcon, TreeItem, TreeItemCollapsibleState, TypeHierarchyItem, UIKind, Uri, ViewColumn, WorkspaceEdit, InlayHint, InlayHintLabelPart, InlayHintKind, NotebookRange, NotebookCellKind, NotebookCellData, NotebookData, NotebookCellStatusBarAlignment, NotebookCellOutput, NotebookCellOutputItem, NotebookCellStatusBarItem, NotebookControllerAffinity, LinkedEditingRanges, TestRunRequest, TestMessage, TestTag, TestRunProfileKind, DataTransfer, DataTransferItem, LanguageStatusSeverity, QuickPickItemKind, TabInputText, TabInputTextDiff, TabInputCustom, TabInputNotebook, TabInputNotebookDiff, TabInputWebview, TabInputTerminal, InputBoxValidationSeverity, InlineCompletionList, InlineCompletionTriggerKind, InlineCompletionItem, DocumentDropEdit, NotebookEditorRevealType, SnippetTextEdit, NotebookEdit, LogLevel, TerminalExitReason, CommentThreadState, TelemetryTrustedValue, TextSearchCompleteMessageType, DocumentPasteEdit, ExternalUriOpenerPriority, TreeItemCheckboxState } = api;
export { Breakpoint, CallHierarchyIncomingCall, CallHierarchyItem, CallHierarchyOutgoingCall, CancellationError, CancellationTokenSource, CodeAction, CodeActionKind, CodeActionTriggerKind, CodeLens, Color, ColorInformation, ColorPresentation, ColorThemeKind, CommentMode, CommentThreadCollapsibleState, CommentThreadState, CompletionItem, CompletionItemKind, CompletionItemTag, CompletionList, CompletionTriggerKind, ConfigurationTarget, CustomExecution, DataTransfer, DataTransferItem, DebugAdapterExecutable, DebugAdapterInlineImplementation, DebugAdapterNamedPipeServer, DebugAdapterServer, DebugConfigurationProviderTriggerKind, DebugConsoleMode, DecorationRangeBehavior, Diagnostic, DiagnosticRelatedInformation, DiagnosticSeverity, DiagnosticTag, Disposable, DocumentDropEdit, DocumentHighlight, DocumentHighlightKind, DocumentLink, DocumentPasteEdit, DocumentSymbol, EndOfLine, EnvironmentVariableMutatorType, EvaluatableExpression, EventEmitter, ExtensionKind, ExtensionMode, ExternalUriOpenerPriority, FileChangeType, FileDecoration, FilePermission, FileSystemError, FileType, FoldingRange, FoldingRangeKind, FunctionBreakpoint, Hover, IndentAction, InlayHint, InlayHintKind, InlayHintLabelPart, InlineCompletionItem, InlineCompletionList, InlineCompletionTriggerKind, InlineValueEvaluatableExpression, InlineValueText, InlineValueVariableLookup, InputBoxValidationSeverity, LanguageStatusSeverity, LinkedEditingRanges, Location, LogLevel, MarkdownString, NotebookCellData, NotebookCellKind, NotebookCellOutput, NotebookCellOutputItem, NotebookCellStatusBarAlignment, NotebookCellStatusBarItem, NotebookControllerAffinity, NotebookData, NotebookEdit, NotebookEditorRevealType, NotebookRange, OverviewRulerLane, ParameterInformation, Position, ProcessExecution, ProgressLocation, QuickInputButtons, QuickPickItemKind, Range, RelativePattern, Selection, SelectionRange, SemanticTokens, SemanticTokensBuilder, SemanticTokensEdit, SemanticTokensEdits, SemanticTokensLegend, ShellExecution, ShellQuoting, SignatureHelp, SignatureHelpTriggerKind, SignatureInformation, SnippetString, SnippetTextEdit, SourceBreakpoint, StatusBarAlignment, SymbolInformation, SymbolKind, SymbolTag, TabInputCustom, TabInputNotebook, TabInputNotebookDiff, TabInputTerminal, TabInputText, TabInputTextDiff, TabInputWebview, Task, TaskGroup, TaskPanelKind, TaskRevealKind, TaskScope, TelemetryTrustedValue, TerminalExitReason, TerminalLink, TerminalLocation, TerminalProfile, TestMessage, TestRunProfileKind, TestRunRequest, TestTag, TextDocumentChangeReason, TextDocumentSaveReason, TextEdit, TextEditorCursorStyle, TextEditorLineNumbersStyle, TextEditorRevealType, TextEditorSelectionChangeKind, TextSearchCompleteMessageType, ThemeColor, ThemeIcon, TreeItem, TreeItemCheckboxState, TreeItemCollapsibleState, TypeHierarchyItem, UIKind, Uri, ViewColumn, WorkspaceEdit, authentication, commands, comments, debug, env, extensions, l10n, languages, notebooks, scm, setDefaultApi, tasks, tests, version, window, workspace };
