// vscode/dist/services.js
export { default as Severity } from 'monaco-editor/esm/vs/base/common/severity.js';
export { StorageScope, StorageTarget } from './vscode/vs/platform/storage/common/storage.js';
import { StandaloneServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices.js';
export { StandaloneServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices.js';
import getServiceOverride from './service-override/layout.js';
import getServiceOverride$1 from './service-override/environment.js';
import getServiceOverride$2 from './service-override/extensions.js';
import getServiceOverride$3 from './service-override/files.js';
import getServiceOverride$4 from './service-override/quickaccess.js';
import { startup, serviceInitializedBarrier } from './lifecycle.js';
export { SyncDescriptor } from 'monaco-editor/esm/vs/platform/instantiation/common/descriptors.js';
export { ICommandService } from 'monaco-editor/esm/vs/platform/commands/common/commands.js';
export { INotificationService, NeverShowAgainScope, NoOpNotification, NoOpProgress, NotificationPriority, NotificationsFilter } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
export { IBulkEditService } from 'monaco-editor/esm/vs/editor/browser/services/bulkEditService.js';
export { ICodeEditorService } from 'monaco-editor/esm/vs/editor/browser/services/codeEditorService.js';
export { ILanguageService } from 'monaco-editor/esm/vs/editor/common/languages/language.js';
export { ILanguageConfigurationService } from 'monaco-editor/esm/vs/editor/common/languages/languageConfigurationRegistry.js';
export { IEditorWorkerService } from 'monaco-editor/esm/vs/editor/common/services/editorWorker.js';
export { ILanguageFeaturesService } from 'monaco-editor/esm/vs/editor/common/services/languageFeatures.js';
export { IModelService } from 'monaco-editor/esm/vs/editor/common/services/model.js';
export { ITextModelService } from 'monaco-editor/esm/vs/editor/common/services/resolverService.js';
export { IClipboardService } from 'monaco-editor/esm/vs/platform/clipboard/common/clipboardService.js';
import './override/vs/platform/dialogs/common/dialogs.js';
export { IFileService } from 'monaco-editor/esm/vs/platform/files/common/files.js';
export { IInstantiationService } from 'monaco-editor/esm/vs/platform/instantiation/common/instantiation.js';
export { IMarkerService } from 'monaco-editor/esm/vs/platform/markers/common/markers.js';
export { IOpenerService } from 'monaco-editor/esm/vs/platform/opener/common/opener.js';
export { IProductService } from 'monaco-editor/esm/vs/platform/product/common/productService.js';
export { IQuickInputService } from 'monaco-editor/esm/vs/platform/quickinput/common/quickInput.js';
export { ITelemetryService } from 'monaco-editor/esm/vs/platform/telemetry/common/telemetry.js';
export { IUriIdentityService } from 'monaco-editor/esm/vs/platform/uriIdentity/common/uriIdentity.js';
export { IBreadcrumbsService } from './vscode/vs/workbench/browser/parts/editor/breadcrumbs.js';
export { GroupDirection, GroupLocation, GroupOrientation, GroupsArrangement, GroupsOrder, IEditorGroupsService, MergeGroupMode } from './vscode/vs/workbench/services/editor/common/editorGroupsService.js';
export { CloseDirection, EditorsOrder } from './vscode/vs/workbench/common/editor.js';
export { EditorInput } from './vscode/vs/workbench/common/editor/editorInput.js';
export { IEditorService } from './vscode/vs/workbench/services/editor/common/editorService.js';
export { IEditorResolverService } from './vscode/vs/workbench/services/editor/common/editorResolverService.js';
export { ITextEditorService } from './vscode/vs/workbench/services/textfile/common/textEditorService.js';
export { IWorkbenchEnvironmentService } from './vscode/vs/workbench/services/environment/common/environmentService.js';
export { IHostService } from './vscode/vs/workbench/services/host/browser/host.js';
export { ILanguageStatusService } from './vscode/vs/workbench/services/languageStatus/common/languageStatusService.js';
export { IPaneCompositePartService } from './vscode/vs/workbench/services/panecomposite/browser/panecomposite.js';
export { IPathService } from './vscode/vs/workbench/services/path/common/pathService.js';
export { ITextFileService } from './vscode/vs/workbench/services/textfile/common/textfiles.js';
export { IWorkingCopyFileService } from './vscode/vs/workbench/services/workingCopy/common/workingCopyFileService.js';
export { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
export { IContextKeyService } from 'monaco-editor/esm/vs/platform/contextkey/common/contextkey.js';
export { IThemeService } from 'monaco-editor/esm/vs/platform/theme/common/themeService.js';
export { ISnippetsService } from './vscode/vs/workbench/contrib/snippets/browser/snippets.js';
export { IWorkspaceContextService } from 'monaco-editor/esm/vs/platform/workspace/common/workspace.js';
export { IStorageService } from 'monaco-editor/esm/vs/platform/storage/common/storage.js';
export { IAudioCueService } from 'monaco-editor/esm/vs/platform/audioCues/browser/audioCueService.js';
export { IDebugService } from './vscode/vs/workbench/contrib/debug/common/debug.js';
export { ILogService, ILoggerService, LogLevel } from 'monaco-editor/esm/vs/platform/log/common/log.js';
export { IViewDescriptorService, IViewsService } from './vscode/vs/workbench/common/views.js';
export { ILayoutService } from 'monaco-editor/esm/vs/platform/layout/browser/layoutService.js';
export { IPreferencesService } from './vscode/vs/workbench/services/preferences/common/preferences.js';
export { IPreferencesSearchService } from './vscode/vs/workbench/contrib/preferences/common/preferences.js';
export { IKeybindingEditingService } from './vscode/vs/workbench/services/keybinding/common/keybindingEditing.js';
export { IOutputService } from './vscode/vs/workbench/services/output/common/output.js';
export { GoFilter, GoScope, IHistoryService } from './vscode/vs/workbench/services/history/common/history.js';
export { IActivityService } from './vscode/vs/workbench/services/activity/common/activity.js';
export { IHoverService } from './vscode/vs/workbench/services/hover/browser/hover.js';
export { IExplorerService } from './vscode/vs/workbench/contrib/files/browser/files.js';
export { IStatusbarService } from './vscode/vs/workbench/services/statusbar/browser/statusbar.js';
export { IDialogService } from './vscode/vs/platform/dialogs/common/dialogs.js';
async function initialize(overrides) {
    const instantiationService = StandaloneServices.initialize({
        ...getServiceOverride(),
        ...getServiceOverride$1(),
        ...getServiceOverride$2(),
        ...getServiceOverride$3(),
        ...getServiceOverride$4(),
        ...overrides
    });
    await startup(instantiationService);
}
async function getService(identifier) {
    await serviceInitializedBarrier.wait();
    return StandaloneServices.get(identifier);
}
export { getService, initialize };
