// vscode/dist/vscode/vs/workbench/contrib/files/browser/workspaceWatcher.js
import { __decorate, __param } from '../../../../../../node_modules/tslib/tslib.es6.js';
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { Disposable, dispose, DisposableStore } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
import { IWorkspaceContextService } from 'monaco-editor/esm/vs/platform/workspace/common/workspace.js';
import { ResourceMap } from 'monaco-editor/esm/vs/base/common/map.js';
import { Severity, NeverShowAgainScope, NotificationPriority, INotificationService } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
import { IOpenerService } from 'monaco-editor/esm/vs/platform/opener/common/opener.js';
import { isAbsolute } from 'monaco-editor/esm/vs/base/common/path.js';
import { IUriIdentityService } from 'monaco-editor/esm/vs/platform/uriIdentity/common/uriIdentity.js';
import { IHostService } from '../../../services/host/browser/host.js';
import { IWorkbenchFileService } from '../../../services/files/common/files.js';
let WorkspaceWatcher = class WorkspaceWatcher extends Disposable {
    constructor(fileService, configurationService, contextService, notificationService, openerService, uriIdentityService, hostService) {
        super();
        this.fileService = fileService;
        this.configurationService = configurationService;
        this.contextService = contextService;
        this.notificationService = notificationService;
        this.openerService = openerService;
        this.uriIdentityService = uriIdentityService;
        this.hostService = hostService;
        this.watchedWorkspaces = ( new ResourceMap(resource => this.uriIdentityService.extUri.getComparisonKey(resource)));
        this.registerListeners();
        this.refresh();
    }
    registerListeners() {
        this._register(this.contextService.onDidChangeWorkspaceFolders(e => this.onDidChangeWorkspaceFolders(e)));
        this._register(this.contextService.onDidChangeWorkbenchState(() => this.onDidChangeWorkbenchState()));
        this._register(this.configurationService.onDidChangeConfiguration(e => this.onDidChangeConfiguration(e)));
        this._register(this.fileService.onDidWatchError(error => this.onDidWatchError(error)));
    }
    onDidChangeWorkspaceFolders(e) {
        for (const removed of e.removed) {
            this.unwatchWorkspace(removed);
        }
        for (const added of e.added) {
            this.watchWorkspace(added);
        }
    }
    onDidChangeWorkbenchState() {
        this.refresh();
    }
    onDidChangeConfiguration(e) {
        if (e.affectsConfiguration('files.watcherExclude') || e.affectsConfiguration('files.watcherInclude')) {
            this.refresh();
        }
    }
    onDidWatchError(error) {
        const msg = ( error.toString());
        if (msg.indexOf('ENOSPC') >= 0) {
            this.notificationService.prompt(Severity.Warning, ( localize(
                'enospcError',
                "Unable to watch for file changes in this large workspace folder. Please follow the instructions link to resolve this issue."
            )), [{
                    label: ( localize('learnMore', "Instructions")),
                    run: () => this.openerService.open(( URI.parse('https://go.microsoft.com/fwlink/?linkid=867693')))
                }], {
                sticky: true,
                neverShowAgain: { id: 'ignoreEnospcError', isSecondary: true, scope: NeverShowAgainScope.WORKSPACE }
            });
        }
        else if (msg.indexOf('EUNKNOWN') >= 0) {
            this.notificationService.prompt(Severity.Warning, ( localize(
                'eshutdownError',
                "File changes watcher stopped unexpectedly. A reload of the window may enable the watcher again unless the workspace cannot be watched for file changes."
            )), [{
                    label: ( localize('reload', "Reload")),
                    run: () => this.hostService.reload()
                }], {
                sticky: true,
                priority: NotificationPriority.SILENT
            });
        }
    }
    watchWorkspace(workspace) {
        const excludes = [];
        const config = this.configurationService.getValue({ resource: workspace.uri });
        if (config.files?.watcherExclude) {
            for (const key in config.files.watcherExclude) {
                if (config.files.watcherExclude[key] === true) {
                    excludes.push(key);
                }
            }
        }
        const pathsToWatch = ( new ResourceMap(uri => this.uriIdentityService.extUri.getComparisonKey(uri)));
        pathsToWatch.set(workspace.uri, workspace.uri);
        if (config.files?.watcherInclude) {
            for (const includePath of config.files.watcherInclude) {
                if (!includePath) {
                    continue;
                }
                if (isAbsolute(includePath)) {
                    const candidate = URI.file(includePath).with({ scheme: workspace.uri.scheme });
                    if (this.uriIdentityService.extUri.isEqualOrParent(candidate, workspace.uri)) {
                        pathsToWatch.set(candidate, candidate);
                    }
                }
                else {
                    const candidate = workspace.toResource(includePath);
                    pathsToWatch.set(candidate, candidate);
                }
            }
        }
        const disposables = ( new DisposableStore());
        for (const [, pathToWatch] of pathsToWatch) {
            disposables.add(this.fileService.watch(pathToWatch, { recursive: true, excludes }));
        }
        this.watchedWorkspaces.set(workspace.uri, disposables);
    }
    unwatchWorkspace(workspace) {
        if (( this.watchedWorkspaces.has(workspace.uri))) {
            dispose(this.watchedWorkspaces.get(workspace.uri));
            this.watchedWorkspaces.delete(workspace.uri);
        }
    }
    refresh() {
        this.unwatchWorkspaces();
        for (const folder of this.contextService.getWorkspace().folders) {
            this.watchWorkspace(folder);
        }
    }
    unwatchWorkspaces() {
        for (const [, disposable] of this.watchedWorkspaces) {
            disposable.dispose();
        }
        this.watchedWorkspaces.clear();
    }
    dispose() {
        super.dispose();
        this.unwatchWorkspaces();
    }
};
WorkspaceWatcher = ( __decorate([
    ( __param(0, IWorkbenchFileService)),
    ( __param(1, IConfigurationService)),
    ( __param(2, IWorkspaceContextService)),
    ( __param(3, INotificationService)),
    ( __param(4, IOpenerService)),
    ( __param(5, IUriIdentityService)),
    ( __param(6, IHostService))
], WorkspaceWatcher));
export { WorkspaceWatcher };
