// vscode/dist/vscode/vs/workbench/api/browser/mainThreadKeytar.js
import { __decorate, __param } from '../../../../../node_modules/tslib/tslib.es6.js';
import { extHostNamedCustomer } from '../../services/extensions/common/extHostCustomers.js';
import { MainContext } from '../common/extHost.protocol.js';
import { ICredentialsService } from '../../../platform/credentials/common/credentials.js';
let MainThreadKeytar = class MainThreadKeytar {
    constructor(_extHostContext, _credentialsService) {
        this._credentialsService = _credentialsService;
    }
    async $getPassword(service, account) {
        return this._credentialsService.getPassword(service, account);
    }
    async $setPassword(service, account, password) {
        return this._credentialsService.setPassword(service, account, password);
    }
    async $deletePassword(service, account) {
        return this._credentialsService.deletePassword(service, account);
    }
    async $findPassword(service) {
        return this._credentialsService.findPassword(service);
    }
    async $findCredentials(service) {
        return this._credentialsService.findCredentials(service);
    }
    dispose() {
    }
};
MainThreadKeytar = __decorate([
    extHostNamedCustomer(MainContext.MainThreadKeytar),
    ( __param(1, ICredentialsService))
], MainThreadKeytar);
export { MainThreadKeytar };
