// vscode/dist/vscode/vs/workbench/browser/parts/statusbar/statusbarItem.js
import { __decorate, __param } from '../../../../../../node_modules/tslib/tslib.es6.js';
import { toErrorMessage } from 'monaco-editor/esm/vs/base/common/errorMessage.js';
import { Disposable, MutableDisposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { SimpleIconLabel } from '../../../../base/browser/ui/iconLabel/simpleIconLabel.js';
import { ICommandService } from 'monaco-editor/esm/vs/platform/commands/common/commands.js';
import { ITelemetryService } from 'monaco-editor/esm/vs/platform/telemetry/common/telemetry.js';
import { ShowTooltipCommand } from '../../../services/statusbar/browser/statusbar.js';
import { IThemeService } from 'monaco-editor/esm/vs/platform/theme/common/themeService.js';
import { isThemeColor } from 'monaco-editor/esm/vs/editor/common/editorCommon.js';
import { show, hide, addDisposableListener, EventType, EventHelper, append } from 'monaco-editor/esm/vs/base/browser/dom.js';
import { INotificationService } from 'monaco-editor/esm/vs/platform/notification/common/notification.js';
import { assertIsDefined } from 'monaco-editor/esm/vs/base/common/types.js';
import { StandardKeyboardEvent } from 'monaco-editor/esm/vs/base/browser/keyboardEvent.js';
import { renderIcon, renderLabelWithIcons } from 'monaco-editor/esm/vs/base/browser/ui/iconLabel/iconLabels.js';
import { syncing, spinningLoading } from 'monaco-editor/esm/vs/platform/theme/common/iconRegistry.js';
import { setupCustomHover } from 'monaco-editor/esm/vs/base/browser/ui/iconLabel/iconLabelHover.js';
import { isMarkdownString, markdownStringEqual } from 'monaco-editor/esm/vs/base/common/htmlContent.js';
import { Gesture, EventType as EventType$1 } from 'monaco-editor/esm/vs/base/browser/touch.js';
let StatusbarEntryItem = class StatusbarEntryItem extends Disposable {
    get name() {
        return assertIsDefined(this.entry).name;
    }
    get hasCommand() {
        return typeof this.entry?.command !== 'undefined';
    }
    constructor(container, entry, hoverDelegate, commandService, notificationService, telemetryService, themeService) {
        super();
        this.container = container;
        this.hoverDelegate = hoverDelegate;
        this.commandService = commandService;
        this.notificationService = notificationService;
        this.telemetryService = telemetryService;
        this.themeService = themeService;
        this.entry = undefined;
        this.foregroundListener = this._register(( new MutableDisposable()));
        this.backgroundListener = this._register(( new MutableDisposable()));
        this.commandMouseListener = this._register(( new MutableDisposable()));
        this.commandTouchListener = this._register(( new MutableDisposable()));
        this.commandKeyboardListener = this._register(( new MutableDisposable()));
        this.hover = undefined;
        this.labelContainer = document.createElement('a');
        this.labelContainer.tabIndex = -1;
        this.labelContainer.setAttribute('role', 'button');
        this.labelContainer.className = 'statusbar-item-label';
        this._register(Gesture.addTarget(this.labelContainer));
        this.label = ( new StatusBarCodiconLabel(this.labelContainer));
        this.container.appendChild(this.labelContainer);
        this.beakContainer = document.createElement('div');
        this.beakContainer.className = 'status-bar-item-beak-container';
        this.container.appendChild(this.beakContainer);
        this.update(entry);
    }
    update(entry) {
        this.label.showProgress = entry.showProgress ?? false;
        if (!this.entry || entry.text !== this.entry.text) {
            this.label.text = entry.text;
            if (entry.text) {
                show(this.labelContainer);
            }
            else {
                hide(this.labelContainer);
            }
        }
        if (!this.entry || entry.ariaLabel !== this.entry.ariaLabel) {
            this.container.setAttribute('aria-label', entry.ariaLabel);
            this.labelContainer.setAttribute('aria-label', entry.ariaLabel);
        }
        if (!this.entry || entry.role !== this.entry.role) {
            this.labelContainer.setAttribute('role', entry.role || 'button');
        }
        if (!this.entry || !this.isEqualTooltip(this.entry, entry)) {
            const hoverContents = isMarkdownString(entry.tooltip) ? { markdown: entry.tooltip, markdownNotSupportedFallback: undefined } : entry.tooltip;
            if (this.hover) {
                this.hover.update(hoverContents);
            }
            else {
                this.hover = this._register(setupCustomHover(this.hoverDelegate, this.container, hoverContents));
            }
        }
        if (!this.entry || entry.command !== this.entry.command) {
            this.commandMouseListener.clear();
            this.commandTouchListener.clear();
            this.commandKeyboardListener.clear();
            const command = entry.command;
            if (command && (command !== ShowTooltipCommand || this.hover) ) {
                this.commandMouseListener.value = addDisposableListener(this.labelContainer, EventType.CLICK, () => this.executeCommand(command));
                this.commandTouchListener.value = addDisposableListener(this.labelContainer, EventType$1.Tap, () => this.executeCommand(command));
                this.commandKeyboardListener.value = addDisposableListener(this.labelContainer, EventType.KEY_DOWN, e => {
                    const event = ( new StandardKeyboardEvent(e));
                    if (event.equals(10 ) || event.equals(3 )) {
                        EventHelper.stop(e);
                        this.executeCommand(command);
                    }
                });
                this.labelContainer.classList.remove('disabled');
            }
            else {
                this.labelContainer.classList.add('disabled');
            }
        }
        if (!this.entry || entry.showBeak !== this.entry.showBeak) {
            if (entry.showBeak) {
                this.container.classList.add('has-beak');
            }
            else {
                this.container.classList.remove('has-beak');
            }
        }
        if (!this.entry || entry.color !== this.entry.color) {
            this.applyColor(this.labelContainer, entry.color);
        }
        if (!this.entry || entry.backgroundColor !== this.entry.backgroundColor) {
            this.container.classList.toggle('has-background-color', !!entry.backgroundColor);
            this.applyColor(this.container, entry.backgroundColor, true);
        }
        this.entry = entry;
    }
    isEqualTooltip({ tooltip }, { tooltip: otherTooltip }) {
        if (tooltip === undefined) {
            return otherTooltip === undefined;
        }
        if (isMarkdownString(tooltip)) {
            return isMarkdownString(otherTooltip) && markdownStringEqual(tooltip, otherTooltip);
        }
        return tooltip === otherTooltip;
    }
    async executeCommand(command) {
        if (command === ShowTooltipCommand) {
            this.hover?.show(true );
        }
        else {
            const id = typeof command === 'string' ? command : command.id;
            const args = typeof command === 'string' ? [] : command.arguments ?? [];
            this.telemetryService.publicLog2('workbenchActionExecuted', { id, from: 'status bar' });
            try {
                await this.commandService.executeCommand(id, ...args);
            }
            catch (error) {
                this.notificationService.error(toErrorMessage(error));
            }
        }
    }
    applyColor(container, color, isBackground) {
        let colorResult = undefined;
        if (isBackground) {
            this.backgroundListener.clear();
        }
        else {
            this.foregroundListener.clear();
        }
        if (color) {
            if (isThemeColor(color)) {
                colorResult = this.themeService.getColorTheme().getColor(color.id)?.toString();
                const listener = this.themeService.onDidColorThemeChange(theme => {
                    const colorValue = theme.getColor(color.id)?.toString();
                    if (isBackground) {
                        container.style.backgroundColor = colorValue ?? '';
                    }
                    else {
                        container.style.color = colorValue ?? '';
                    }
                });
                if (isBackground) {
                    this.backgroundListener.value = listener;
                }
                else {
                    this.foregroundListener.value = listener;
                }
            }
            else {
                colorResult = color;
            }
        }
        if (isBackground) {
            container.style.backgroundColor = colorResult ?? '';
        }
        else {
            container.style.color = colorResult ?? '';
        }
    }
};
StatusbarEntryItem = ( __decorate([
    ( __param(3, ICommandService)),
    ( __param(4, INotificationService)),
    ( __param(5, ITelemetryService)),
    ( __param(6, IThemeService))
], StatusbarEntryItem));
class StatusBarCodiconLabel extends SimpleIconLabel {
    constructor(container) {
        super(container);
        this.container = container;
        this.progressCodicon = renderIcon(syncing);
        this.currentText = '';
        this.currentShowProgress = false;
    }
    set showProgress(showProgress) {
        if (this.currentShowProgress !== showProgress) {
            this.currentShowProgress = showProgress;
            this.progressCodicon = renderIcon(showProgress === 'loading' ? spinningLoading : syncing);
            this.text = this.currentText;
        }
    }
    set text(text) {
        if (this.currentShowProgress) {
            if (this.container.firstChild !== this.progressCodicon) {
                this.container.appendChild(this.progressCodicon);
            }
            for (const node of Array.from(this.container.childNodes)) {
                if (node !== this.progressCodicon) {
                    node.remove();
                }
            }
            let textContent = text ?? '';
            if (textContent) {
                textContent = ` ${textContent}`;
            }
            append(this.container, ...renderLabelWithIcons(textContent));
        }
        else {
            super.text = text;
        }
    }
}
export { StatusbarEntryItem };
